### ISSUE TITLE
Allow External Authentication Providers

### ISSUE THREAD

**Issue by DavidMStraub** [gramps-project/gramps-web-api#355]
https://flask-dance.readthedocs.io/en/latest/

**Comment by cdhorn** [comment]
May want to consider https://python-social-auth.readthedocs.io/en/latest/intro.html#features as well. Ansible Tower uses it.

**Comment by cypherkey** [comment]
Hello.

I've begun working a little bit on this.  Unfortunately all my experience is with backend work and I have little to no experience on SPA or front-end Javascript.  

Using this [site](https://realpython.com/flask-google-login/), I created API endpoints (login and callback) for Google. The callback endpoint then creates a JWT just like the token endpoint does. The code can be found [here](https://pastebin.com/xgYQkB0L). **NOTE**: The code is rough and could be cleaned up. There is also duplication of code from token.py.

The issue I have is how to tie this to the front-end. Do you have any thoughts on how you would want this implemented?  The alternative that I can look in to implementing SSO in the JS front end.

**Comment by DavidMStraub** [comment]
Thanks for looking into this, but I suggest to first think about architectural issues before starting to code. When I looked into this myself, I stopped at some point because it was getting too complicated for me. There are many issues to solve, like where to store user assocation between the internal users and the SSO providers, and how toi handle signup; see https://flask-dance.readthedocs.io/en/latest/multi-user.html for some discussion (whether we use Flask-Dance or not).

In your code, only emails are compared, but is this enough? If the user changes the email address associated with the SSO provider, they will completely loose access. (Probably this is unlikely for Google, but not for other providers.)

**Comment by PovilasID** [comment]
I suggest looking into https://pypi.org/project/oic/ it is supported by things like Keycloak which has the widest compatibility.  https://flask-pyoidc.readthedocs.io/en/stable/quickstart.html looks like a good option too. Keycloak can plug other providers like Google into it or use it directly because most have OIDC integration but mostly Keycloak makes sense for development because it will throw errors on "provider" side that you can check.
I think providers Google; facebook etc. will define what prams are needed (email, id, etc) for things like Keycloak you can define required and minimal prams. 
Here is something that may be useful: https://gist.github.com/thomasdarimont/145dc9aa857b831ff2eff221b79d179a
Users using 3rd party services often include 'openid' as and additional user ID. This permits using local and 3rd party logins in parallel.

**Comment by AquaWolf** [comment]
Does someone already started with this issue? I would like to help somehow. I'm more a frontend developer. If you already considered a framework I could dig a bit deeper.

**Comment by DavidMStraub** [comment]
Hi, no, nobody is working on it right now - unfortunately, I think it is mostly a backend challenge.

**Comment by shotor** [comment]
@DavidMStraub can I work on this? I have experience with it and I've been wanting to run gramps for a while but need this feature :)

There are 2 common ways to do account matching:

Most common: As mentioned, just use the email and store the link. After the link has been established, you can change the email in the user account or oauth account and it won't have effect. So you'll still be able to login with your oauth account. Remove the link, and you lose access. In this case you can still let the account fallback to a normal account, and a password reset will get your account back. Provided you still have access to the email of the *user account* (not oauth account). Most apps provide external login like this and I think it's an acceptable risk for most apps.

More complicated: On initial login with an oauth account, so no account links exist yet. Ask the user if they want to create a new account, or link to an existing account. If they choose the latter, have them do a user/password login, then link the account. Now local accounts and oauth accounts are decoupled. Apps that need to provide more resilience use this. For instance this is how Gitea does it. You don't see this a lot though.

Next thing is to choose what you want to actually provide: OIDC, OAuth2, SAML, LDAP.

I would focus on OIDC. Keeps the code simple, most popular providers support it. And if they don't, they should.

Storing the links between users and external logins, 2 common ways I see: Single link; a user account can be associated. Or multiple links; a user account can be associated with multiple external logins. I prefer the latter, there's not much of a difference in implementation and provides some resiliency if you lose access to an account.

You'll sometimes see apps disabling the password if there's an external login configured. I think that's a good idea. Since your external provider probably provides you with 2factor.

**Comment by DavidMStraub** [comment]
Yes it would be awesome if you work on this, and since you understand the topic much better than me, I trust your judgement. Let me know if you need guidance to link everything up with Gramps Web API. Concerning the database table, it should be as simple as adding the relevant SQLAlchemy models and creating an alembic migration script for the user database.

One of the things I haven't grasped yet: what is the workflow for new users to sign up? When choosing the first (less complicated, which is good) option, is the registration & first login still as it is now, and users will only get a button in their profile (just thinking already about the frontend) to link to an existing external account? If so, what is actually the main benefit? Added security?

**Comment by shotor** [comment]
When trying to sign up, the user  clicks sign up with "the configured oidc provider". After the oauth flow the backend will have access to the users name, email etc. through the token provided by the identity provider (idp). It can then create an account based on that, instead of having the user fill those fields in. You can even configure user permissions through roles or scopes provided by the idp provider.

You can in addition add a "add new oidc account" to the user profile page, but is not strictly necessary. When you try to login or register with the oidc account and the emails match the link is automatically created.

The main benefit is integration with your own infrastructure. I host a keycloak server and a bunch of apps for my friends and family. They can login to all those apps with the same account. I think this is why most people want this feature.

In addition I can centrally manage the security through keycloak: I require passkey (webauthn) 2fa for all accounts. But this can be added to gramps too without an idp provider.

**Comment by DavidMStraub** [comment]
Ok that sounds perfect!

**Comment by HytrexOfficial** [comment]
@shotor Sorry for asking, but do you have any timeline for this feature? 

This is the only reason why i haven't published gramps to the family.

Unfortunately i can't support this topic/the project with my techstack :/

**Comment by Allexio** [comment]
I've got nothing against passwordless via email logins as a feature but I don't see how it would accomplish what we want from oidc.
I would not be using that feature and as an administrator I would probably not enable it for my users.
I feel like the kind of data hosted on a Gramps instance is very sensitive, so having strong authentication for stuff like this is paramount.

**Comment by Allexio** [comment]
Keep in mind this is only my opinion and I probably don't speak for everyone, especially the less tech savvy users who may not know how to set up an oidc platform or simply don't know if their existence.

If immich (open source photo hosting service similar to Google photos) had baked in 2FA, I would have never found out about self hosted authentication providers such as Authelia or keycloak.

**Comment by DavidMStraub** [comment]
Interesting, I didn't even know immich, I use photoprism and it apparently supports OIDC https://docs.photoprism.app/developer-guide/api/oidc/ 🙂  might help as inspiration for our implementation.

**Comment by Allexio** [comment]
Also just an extra two cents on the email feature : this may be easy to set up on your side but may require significant extra steps on system admin side when it comes to actually providing a way to send emails.

**Comment by Allexio** [comment]
(sorry for the off topicness)
I feel like my side note wasn't very clear: I had no idea what OIDC was or that open source authentication providers were a thing.

I wanted MFA on immich but immich did not support MFA, immich did however support OIDC so that is how I dug deeper and found authentication providers and deployed my own and I do not regret it at all.

It's actually really easy to set up, there's multiple options and they all come with extensive documentation and healthy communities of people who are there to help you get set up.

Point is that I think the biggest roadblock to more Auth provider adoption is probably not the complexity of the setup but rather the initial knowledge of their existence and use 😅

**Comment by DavidMStraub** [comment]
This might be the right lib to use, what do you think @shotor? https://docs.authlib.org/en/latest/client/flask.html

**Comment by HytrexOfficial** [comment]
> > This is the only reason why i haven't published gramps to the family.
> 
> Just to understand better, why is this blocking you from using Gramps Web? Because you don't trust them to use secure passwords or because it's too inconvenient?


You said it, i don't trust them that they use secure passwords. And im currently setting up Zitadel as an identity provider where i can force a second factor or passkey and so on or that they can use their google, Microsoft whatever account. Benefit here is then also that i can provide them multiple services with only this single account they need to remember.

On top of that is that i have then a better overview over the users and security in general. 

For me is gramps a application with highly sensitive data which i want to protect as best from persons outside of the family tree. 

I use gramps and gramps Web already only reachable for me. 

The option with the email login is also nice and for not so experienced people in the infrastructure sector definitely a simple way for a fast and secure login


Here a shout out to all people who contribute to these projects. The application and the Web Version are really nice and definitely better then all the online trees, especially because of the self hosted possibility.

**Comment by Enrico204** [comment]
I would like to add myself to the list of happy users who are praising this project - kudos to all developers (Gramps and Gramps Web alike!).

On the authentication: I would like to mention a potential "alternative" to OIDC: authenticate via headers. It should be way easier than OIDC, and it is still possible to use OIDC to implement that -- for example, by putting `oauth2-proxy` in front of Gramps Web.

In terms of deployments, I see two possible alternatives:

1. A reverse proxy with support for "forward authentication" to an authentication provider (e.g., Caddy + Authelia): `Ingress -> Caddy -> Authelia` for the authentication, then this automatically changes into `Ingress -> Caddy -> Gramps Web` . See <https://caddyserver.com/docs/caddyfile/directives/forward_auth> for Caddy. This is supported by some reverse proxies.
2. `oauth2-proxy` as reverse proxy (it might be "after" other reverse proxies): `Ingress -> oauth2-proxy -> Gramps Web`. In this case, `oauth2-proxy` handles the OIDC/OAuth2 flows.

In both cases, Gramps Web should only support reading one (or more) headers with the relevant information, and trust them.

(support for LDAP, SAML, etc. is already present in many identity providers, such as Authelia and Keycloak, so there is no need to reinvent the wheel here).

**Comment by DavidMStraub** [comment]
Since there is on the one hand significant user interest in this feature (referring to OIDC now), and on the other hand this is a significant chunk of work, I think it's a good candidate for using some of the funding that's available thanks to [Grampshub](https://www.grampshub.com/). Not in the sense of fully remunerating the development work, but in the sense of a bounty as incentive. If anyone is interested and feels up to the challenge, please contact me via support@grampshub.atlassian.net and we can work out the details.

**Comment by DavidMStraub** [comment]
(No offense @DestruktServices, interesting ideas, but out of scope for this issue, feel free to open a discussion thread.)

**Comment by Allexio** [comment]
> Since there is on the one hand significant user interest in this feature (referring to OIDC now), and on the other hand this is a significant chunk of work, I think it's a good candidate for using some of the funding that's available thanks to [Grampshub](https://www.grampshub.com/). Not in the sense of fully remunerating the development work, but in the sense of a bounty as incentive. If anyone is interested and feels up to the challenge, please contact me via [support@grampshub.atlassian.net](mailto:support@grampshub.atlassian.net) and we can work out the details.

Is there a way to add to this bounty for specifically this feature?
Like if I want to give 50 bucks to the guy who develops this?

**Comment by DavidMStraub** [comment]
I haven't set up any bounty service (open to suggestions), but I'm sure we'll find a way to do it.

**Comment by Allexio** [comment]
I think for something like this there should also be clear requirements to get the bounty, because otherwise it might be confusing what has to be done to get it.

Basic features that should be required for minimal usage of OIDC:
- Implement OIDC support for Gramps Web
- Add a button in the Gramps Web login UI to "log in with OIDC"

QoL features that should definitely be considered and that I would very much appreciate:
- Allow admins to disable Gramps Web login (to only enable login via the authentication platform)
- Allow admins to enable auto-redirect to the authentication service when attempting to access Gramps Web
- Implement group handling so groups/roles can be passed from the authentication service to Gramps Web to enable easy rights management


As for the bounty platform, I can't say I'm aware of any that has this level of granularity. When I support a project I generally donate to the project itself, not a specific feature of that project 😅

**Comment by DavidMStraub** [comment]
I agree 100% clear requirements are needed - that's actually why I didn't set up any of the bounty bots and just provided an e-mail address so we can negotiate the bounty & requirements - for instance, it would be conceivable that somebody would only implement the backend part rather than frontend + backend, for a reduced bounty. Alas, so far nobody has reached out 🙂

**Comment by JohanAlbrectsen** [comment]
@DavidMStraub 
Would easily donate $100 to a bounty program that solves this issue. It's the last out of my 9 self hosted applications that's missing this feature. Thanks for your great work otherwise 🙏

**Comment by DavidMStraub** [comment]
I already mentioned there's funds available, see https://github.com/gramps-project/gramps-web-api/issues/355#issuecomment-2877449757. Didn't help so far.

**Comment by AlexBocken** [comment]
Always happy to collect the bounty if you consider my implementation good enough. (see #691 and https://github.com/gramps-project/gramps-web/pull/763 ) and Having read through this discussion it seems like I've hit all the major points mentioned here.
If you deem anything still missing feel free to reach out.

**Comment by Allexio** [comment]
> Always happy to collect the bounty if you consider my implementation good enough.

I looked at the PRs briefly and from a feature standpoint, this is exactly what I was looking for.
If it gets successfully reviewed & merged by David, I'd be happy to send you an extra 50 bucks on top of what's in the planned bounty.

**Comment by Allexio** [comment]
@DavidMStraub Would you be able to say when this feature will be accessible in the docker release ?

(also @AlexBocken do you have a donation link or something?)
